Sail Separate integration docs added (#210)
stancl/tenancy-docs: source/docs/v3/integrations/sail.blade.md

---
title: Laravel Sail integration
extends: _layouts.documentation
section: content
---

# Laravel Sail {#sail}

> ## Creating tenants

Reviewer: Heading can be removed, it doesn't add anything here and would lead to confusion in the search.

Author: I removed the `h1`, should I remove the `h2` as well?

Reviewer: I commented on line 9, not line 7. Obviously we want to keep the name of the page. I was referring to how "Creating tenants" would lead to confusing search results.

Author: Sorry, my bad. Fixed this one.

By default sail has the `ALL` privilege on the default database (here it is the central database).
In order to create, read, update and delete operations related to the database, we will assign `ALL` privilege on all databases via the wildcard `*`;

> To do so either login to the mysql shell or via any client using root user credentials and run the following statements

Reviewer: @lukinovec can you rewrite this to improve the grammar a bit. You don't need to keep all of the paragraphs, feel free to make it as short as it needs to be.

Reviewer 2: ```suggestion
Sail can only perform the create, read, update and delete operations in the central database by default. To give Sail permission also to perform these operations in the tenant databases, log in to Sail's MySQL shell as the root user (by default, the username is `root` and the password is `password`) and assign the `ALL` privilege to all the databases by running the following statements:
```

The original text also didn't provide clear info, I think this suggestion should correct that. Originally, in the first paragraph, it's said that Sail has the privilege in the default DB, but in the next, "in order to do the operations related to the database", we assign the privilege to all databases, which is confusing.

Reviewer: @lukinovec I edited your suggestion, let me know your thoughts. And yeah, including that command would be useful since that's what most people will use.

Reviewer 2: The edited suggestion looks good. I edited the parentheses after `log in to Sail's MySQL shell as the root user` to include the commands. We can commit the suggestion if you agree with the change @stancl

---
title: Laravel Sail integration
extends: _layouts.documentation
section: content
---

# Laravel Sail {#sail}

## Creating tenants

> Before creating any tenants you have to ensure that the `sail` user has the necessary permissions.

Reviewer: `the necessary permissions` doesn't make sense here, it should say specifically which permissions. Before telling people what to change, we should tell them why they are doing it — i.e. what permissions are missing.

Author: I have re-framed the statement. Please check how it sounds.

> To do so either login to the mysql shell or via any client using root user credentials and run the following queries:

Reviewer: ```suggestion
To do so, login to the mysql instance using root user credentials and run the following queries:
```

Author: statements are perfect. Replaced.

```bash
GRANT ALL PRIVILEGES on *.* to 'sail'@'%';
FLUSH PRIVILEGES;
```

### Specifying Database Credentials {#specifying-database-credentials}

While working with sail, it offers a database user `sail` which just have the privileges for just the one database created at the inception of the project (central database in our tenancy case). This user does not have the privileges to create new users or `GRANT OPTION` privilege, which is used to grant permissions to other users.

If you want to use custom users (one user for each respective database), you will need `GRANT OPTION` grant on your central database user (`sail` in this case), otherwise you cannot grant permissions to the newly created user and will result in an error.

To avoid this failure, run the following SQL from root user:

```sql
GRANT GRANT OPTION on central_database.* to 'sail'@'%';
FLUSH PRIVILEGES;

> ```

Reviewer: I don't understand the point of this section. This looks like a reiteration of the first section?

Reviewer: You *can* have one user per tenant, the package has a feature for that, but I don't see what this section does differently compared to the previous section.

Author: I was speaking of permissions, specifically this `GRANT OPTION`. Okay, ideally this is not the right place then. Sail has nothing to do with this grant. I will just remove this section then.

---
title: Laravel Sail integration
extends: _layouts.documentation
section: content
---

# Laravel Sail {#sail}

## Central domains

If you're using Laravel Sail, no changes are needed, default values are good to go:

```php
// config/tenancy.php

'central_domains' => [
    '127.0.0.1',
    'localhost',
],

> ```

Reviewer: What's this section for? It made sense in the quickstart, but if this is an integration guide, it shouldn't mention things that **don't** need to be done.

Author: FIxed

## Creating tenants

Before creating any tenants you have to ensure that the `sail` user has the necessary permissions.

To do so either login to the mysql shell or via any client using root user credentials and run the following queries:

> ```bash

Reviewer: Both of these code blocks should be `sql`, not `bash`

Author: Yes I know, I intentionally used `bash` because of styling issue.

![image](https://user-images.githubusercontent.com/70144052/193395395-792d6534-0a96-4a6f-bc3a-f8beb1c2ce76.png)

The first one is `bash`, and the second one is `sql`. Can someone look into the CSS part?

Reviewer: I see, that's probably because we don't have sql in the languages for our code highlighter. `bash` is fine then

Reviewer: Seems like you change this to `sql` but didn't update the styles? So now it looks like on the screenshot above https://6345ccd6c6181c000829be1d--stancl-tenancy.netlify.app/docs/v3/integrations/sail/

Author: Okay, I must have not committed the assets. I should run `npm run prod`, and then commit right?

Author: Fixed

GRANT ALL PRIVILEGES on *.* to 'sail'@'%';
FLUSH PRIVILEGES;
```

> If you are using [Database Customization]({{ $page->link('customizing-databases') }}), then you will have add the following GRANT as well:

Reviewer: Could you expand on this?

Author: Is this fine?

Reviewer: I'm asking you to explain the point of this section

Author: Please check it once again. I have tried to elaborate the section a little more.

Author: Removed the irrelevant part
